### PR TITLE
[Flow EVM] Run tracing hooks `OnTxStart`, `OnTxEnd` for `DryRunTransaction`

### PR DESCRIPTION
**Pull request by m-Peter**
work towards: https://github.com/onflow/flow-evm-gateway/issues/530

This will allow us to generate traces for calls, to be used for `debug_traceCall`.

### CODE REVIEW

File: fvm/evm/emulator/emulator.go [hunk 1]
```diff
@@ -306,6 +311,13 @@ func (bl *BlockView) DryRunTransaction(
 		txResult.GasConsumed += txResult.GasRefund
 	}
 
+	// call tracer on tx end
```
Comment (m-Peter) [inline fvm/evm/emulator/emulator.go:314]: I am not sure if this should be before or after adjusting the `txResult.GasConsumed` with the various buffers.

File: fvm/evm/emulator/emulator.go [hunk 2]
```diff
@@ -306,6 +311,13 @@ func (bl *BlockView) DryRunTransaction(
...
+	if proc.evm.Config.Tracer != nil &&
+		proc.evm.Config.Tracer.OnTxEnd != nil &&
+		txResult != nil {
```
Comment (sideninja) [inline fvm/evm/emulator/emulator.go:317]: `txResult != nil` this should always be true

**Review by sideninja — APPROVED** [review]

File: fvm/evm/emulator/emulator.go (hunk 2 above)
Comment (m-Peter) [reply fvm/evm/emulator/emulator.go:317]: I saw that `proc.run` can return `nil` for `txResult` in some cases:
```go
if err != nil {
	return nil, err
}
```
Besides, this is the exact same condition for both `RunTransaction` & `BatchRunTransaction`, so I wanted to be on the safe side.

**Review by ramtinms — APPROVED** [review]
the changes is okey to me, but I don't think we need to deploy it mainnet right away, if we are running the evm emulator offchain for this goal, then we could use a branch with these changes until post spork.

**Comment by m-Peter** [comment]
> the changes is okey to me, but I don't think we need to deploy it mainnet right away, if we are running the evm emulator offchain for this goal, then we could use a branch with these changes until post spork.

@ramtinms I have found a way to achieve this functionality with the approach in: https://github.com/onflow/flow-evm-gateway/pull/567/files#diff-34c0f187c58bb59d36ca39cc631b07ba8f74d7f6886a5d6c6a796d721a928e67R244-R253. 
What I am concerned about is this line:
```go
// call tracer
tracer.OnTxStart(&tracing.VMContext{}, tx, from)
```
I am not entirely sure if I can populate this exactly as `GetVMContext` does:
```go
// GetVMContext provides context about the block being executed as well as state
// to the tracers.
func (evm *EVM) GetVMContext() *tracing.VMContext {
	return &tracing.VMContext{
		Coinbase:    evm.Context.Coinbase,
		BlockNumber: evm.Context.BlockNumber,
		Time:        evm.Context.Time,
		Random:      evm.Context.Random,
		GasPrice:    evm.TxContext.GasPrice,
		ChainConfig: evm.ChainConfig(),
		StateDB:     evm.StateDB,
	}
}
```
Other than that, the approach on the EVM GW seems to work.

**Comment by ramtinms** [comment]
@m-Peter I think we should merge, my comment was for deployment given we never need tracer on the dryrun for onchain operations.

**Review by janezpodhostnik — APPROVED** [review]
Maybe we need to consider using a no-op tracer instead of nil, to get rid of all these checks and make it a bit cleaner. (Just a general comment, not really for this PR)

**Pull request merged**